File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
from heapq import *
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = [0] * 2001
        for num in nums:
            counts[num + 1000] += 1
        buckets = [[] for _ in range(len(nums) + 1)]  # bucket sort
        for num, freq in enumerate(counts):
            buckets[freq].append(num - 1000)
        ret = []
        for i in range(len(buckets) - 1, 0, -1):
            for num in buckets[i]:
                ret.append(num)
                if len(ret) == k:
                    return ret

File: Data Structures & Algorithms/target-sum/submission-2.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }
        if (target > sum || target < -sum) return 0;
        vector<vector<int>> dp(nums.size(), vector<int>(sum * 2 + 1, 0));
        int R = dp.size(), C = dp[0].size();
        dp[0][nums[0] + sum] = 1;
        dp[0][-nums[0] + sum] += 1;
        for (int r = 1; r < R; ++r) {
            for (int c = 0; c < C; ++c) {
                if (c - nums[r] >= 0) {
                    dp[r][c] += dp[r - 1][c - nums[r]];
                }
                if (c + nums[r] < C) {
                    dp[r][c] += dp[r - 1][c + nums[r]];
                }
            }
        }
        return dp[R - 1][target + sum];
    }
};
